### PR TITLE
Add Base.in for rays.jl

### PR DESCRIPTION
**Pull request by dorn-gerhard**

### CODE REVIEW

File: src/primitives/ray.jl [hunk 1]
```diff
@@ -42,3 +42,17 @@ origin(r::Ray) = r.p
...
 """
 direction(r::Ray) = r.v
+
+
+
+function Base.in(p::Point, r::Ray)
+	w = norm(r.v)
+	d = evaluate(Euclidean(), p, Line(r(0), r(1)))
+	if d+w ≈ w
+		return sum((p - r.p) .* r.v) >= 0 # return true if point is in ray
```
Comment (dorn-gerhard) [inline src/primitives/ray.jl]: This line checks whether the point is in the positive direction of the ray.
The rest is basically the same as in Base.in(p::Point, l::Line)

Comment (juliohm) [inline src/primitives/ray.jl]: Incorrect code style, please fix it to follow the style of the project.

Also, I understood that you are just repeating the code of point and line and then doing an additional check to see if the point is on the right side of the ray. In this case keep it high-level:

```julia
p \in Line(r(0), r(1)) && # additional check
```

Also, notice that 0 and 1 in Julia are Int. You need to use `zero(T)` and `one(T)` to get the same coordinate type.

File: src/primitives/ray.jl [hunk 2]
```diff
@@ -42,3 +42,17 @@ origin(r::Ray) = r.p
...
+	else
+		return false
+	end
+end
+
+==(r1::Ray, r2::Ray) = r1.p == r2.p ∈ r1 && r1(1) ∈ r2 && r2(1) ∈ r1
```
Comment (juliohm) [inline src/primitives/ray.jl]: This line is too confusing. Can you please add parenthesis?

File: test/primitives.jl [hunk 3]
```diff
@@ -22,6 +22,21 @@
     @test r(T(1.)) == P2(1,1)
     @test r(T(Inf)) == P2(Inf,Inf)
     @test_throws DomainError(T(-1), "r(t) is not defined for t < 0.") r(T(-1))
+    
+    r1 = Ray((0,0,0), (1,0,0.))
+    r2 = Ray((1,1,1), (1,2,1.))
+    @test (r1 == r2) == false
```
Comment (juliohm) [inline test/primitives.jl]:
```suggestion
    @test r1 != r2
```

File: test/primitives.jl [hunk 4]
```diff
@@ -22,6 +22,21 @@
...
+  
+    r1 = Ray((0,0,0), (1,0,0.))
+    r2 = Ray((1,0,0), (-1,0,0.))
+    @test (r1 == r2) == false
```
Comment (juliohm) [inline test/primitives.jl]:
```suggestion
    @test r1 != r2
```

**Review by juliohm — CHANGES_REQUESTED** [review]
Left a first round of review comments. Please try to follow the code style next time.

File: src/primitives/ray.jl (hunk 2 above)
Comment (dorn-gerhard) [reply src/primitives/ray.jl]: should be more clear now

**Comment by dorn-gerhard** [comment]
I made the requested changes and getting more and more used to the code style.
Is there a reference for the proper code style?

**Comment by juliohm** [comment]
Regarding the code style, it basically consists of indentation with 2
spaces, avoid underscores when possible and descriptive variable names.

Em sex., 6 de mai. de 2022 10:41, dorn-gerhard ***@***.***>
escreveu:

File: src/primitives/ray.jl [hunk 5]
```diff
@@ -42,3 +42,11 @@ origin(r::Ray) = r.p
 The direction of the ray.
 """
 direction(r::Ray) = r.v
+
+
+
+function Base.in(p::Point{Dim, T}, r::Ray{Dim, T}) where {Dim,T}
+	return p ∈ Line(r(zero(T)), r(one(T))) && sum((p - r.p) .* r.v) >= 0 # additional check
+end
```
Comment (juliohm) [inline src/primitives/ray.jl]: Incorrect indentation and tons of unnecessary whitespace. We don't use `return` either. Appreciate if you can comply to the code style of the project.

**Review by juliohm — CHANGES_REQUESTED** [review]

**Review by dorn-gerhard — COMMENTED** [review]
I adopted the code to match the coding style

File: src/primitives/ray.jl [hunk 6]
```diff
@@ -42,3 +42,9 @@ origin(r::Ray) = r.p
 The direction of the ray.
 """
 direction(r::Ray) = r.v
+
+function Base.in(p::Point{Dim, T}, r::Ray{Dim, T}) where {Dim,T}
+  p ∈ Line(r(zero(T)), r(one(T))) && sum((p - r.p) .* r.v) >= 0 # additional check
+end
```
Comment (juliohm) [inline src/primitives/ray.jl]:
```suggestion
Base.in(p::Point, r::Ray) = p ∈ Line(r.p, r.p + r.v) && (p - r.p) ⋅ r.v ≥ 0
```

File: src/primitives/ray.jl [hunk 7]
```diff
@@ -42,3 +42,9 @@ origin(r::Ray) = r.p
...
+
+==(r1::Ray{Dim, T}, r2::Ray{Dim, T}) where {Dim,T} = (r1.p == r2.p) && (r1(one(T)) ∈ r2) && (r2(one(T)) ∈ r1)
```
Comment (juliohm) [inline src/primitives/ray.jl]:
```suggestion
```

Please add this function in a separate PR.

File: test/primitives.jl [hunk 8]
```diff
@@ -22,6 +22,21 @@
...
+    r2 = Ray((1,1,1), (1,2,1.))
+    @test r1 != r2
+  
+    r1 = Ray((0,0,0), (1,0,0.))
+    r2 = Ray((1,0,0), (-1,0,0.))
+    @test r1 != r2
+  
+    r1 = Ray((0,0,0), (1,0,0.))
+    @test r1 == r1
+
```
Comment (juliohm) [inline test/primitives.jl]:
```suggestion
```

Please move these tests to a separate PR.

File: test/primitives.jl [hunk 9]
```diff
@@ -22,6 +22,21 @@
...
+    r1 = Ray((0,0,0), (1,0,0.))
+    r2 = Ray((1,0,0), (-1,0,0.))
+    @test r1 != r2
+  
+    r1 = Ray((0,0,0), (1,0,0.))
+    @test r1 == r1
+    
+    r1 = Ray((0,0,0), (1,1,1.))
+    P1 = Point(3,3,3)
+    @test P1 ∈ r1
```
Comment (juliohm) [inline test/primitives.jl]: We don't use capital letters for objects. Please rename P1 to p and r1 to r. Also, swap the lines because the function p \in r has arguments in the reverse order.

**Review by juliohm — CHANGES_REQUESTED** [review]

**Review by dorn-gerhard — COMMENTED** [review]
required changes implemented

File: src/primitives/ray.jl [hunk 10]
```diff
@@ -42,3 +42,7 @@ origin(r::Ray) = r.p
 The direction of the ray.
 """
 direction(r::Ray) = r.v
+
+Base.in(p::Point, r::Ray) = p ∈ Line(r.p, r.p + r.v) && (p - r.p) ⋅ r.v ≥ 0
+
+==(r1::Ray{Dim, T}, r2::Ray{Dim, T}) where {Dim,T} = (r1.p == r2.p) && (r1(one(T)) ∈ r2) && (r2(one(T)) ∈ r1)
```
Comment (juliohm) [inline src/primitives/ray.jl]:
```suggestion
==(r₁::Ray, r₂::Ray) = (r₁.p == r₂.p) && (r₁.p + r₁.v) ∈ r₂ && (r₂.p + r₂.v) ∈ r₁
```

File: test/primitives.jl [hunk 11]
```diff
@@ -22,6 +22,21 @@
...
+    r1 = Ray((0,0,0), (1,0,0.))
+    r2 = Ray((1,1,1), (1,2,1.))
+    @test r1 != r2
+  
+    r1 = Ray((0,0,0), (1,0,0.))
+    r2 = Ray((1,0,0), (-1,0,0.))
+    @test r1 != r2
+  
+    r1 = Ray((0,0,0), (1,0,0.))
+    @test r1 == r1
```
Comment (juliohm) [inline test/primitives.jl]:
```suggestion
    p = Point(3,3,3.)
    r = Ray((0,0,0), (1,1,1.))
    @test p ∈ r

    r1 = Ray((0,0,0), (1,0,0.))
    r2 = Ray((1,1,1), (1,2,1.))
    @test r1 != r2
  
    r1 = Ray((0,0,0), (1,0,0.))
    r2 = Ray((1,0,0), (-1,0,0.))
    @test r1 != r2
  
    r = Ray((0,0,0), (1,0,0.))
    @test r == r
```

**Comment by dorn-gerhard** [comment]
Sorry for not replying, I see I should use more unicode like \cdot p\_1 etc. to make the code more readible. 
I added two test cases, for Base.in(::Point, ::Ray) to check cases of a point in the negative ray direction and a point outside the extended ray - line.

**Review by dorn-gerhard — COMMENTED** [review]
I added test cases and use more subscripts / unicode characters to improve readability.

File: test/primitives.jl [hunk 12]
```diff
@@ -22,6 +22,26 @@
...
+    r1 = Ray((0,0,0), (1,0,0.))
+    r2 = Ray((1,1,1), (1,2,1.))
+    @test r1 != r2
+  
+    r1 = Ray((0,0,0), (1,0,0.))
+    r2 = Ray((1,0,0), (-1,0,0.))
+    @test r1 != r2
+  
+    r = Ray((0,0,0), (1,0,0.))
+    @test r == r
```
Comment (juliohm) [inline test/primitives.jl]: Notice that all these tests need to be changed to use the T type as the other tests. @dorn-gerhard we are wasting a lot of time reviewing this PR because you are not following the style of the surrounding code. Can you please pay more attention in the next round of reviews?

Comment (dorn-gerhard) [reply test/primitives.jl]: sorry for wasting time - some things were unclear to me, slow learning curve - I should improve.

File: test/primitives.jl [hunk 13]
```diff
@@ -22,6 +22,25 @@
...
+    r1 = Ray(P3(0,0,0), V3(1,0,0))
+    r2 = Ray(P3(1,1,1), V3(1,2,1))
+    @test r1 != r2
+
+    r1 = Ray(P3(0,0,0), V3(1,0,0))
+    r2 = Ray(P3(1,0,0), V3(-1,0,0))
+    @test r1 != r2
+
+    r = Ray(P3(0,0,0), V3(1,0,0))
+    @test r == r
```
Comment (juliohm) [inline test/primitives.jl]: This test is too trivial. Can you create two rays that have different direction vectors but are still the same? That way we can be sure that the equality holds even when the vectors have different length.

**Review by juliohm — CHANGES_REQUESTED** [review]
Attached comments.

File: test/primitives.jl (hunk 13 above)
Comment (dorn-gerhard) [reply test/primitives.jl]: Sure, I added two cases, 
one with varied vector length (true)
one with identical vectors but different origin (false)

**Comment by juliohm** [comment]
Thank you!
